<!DOCTYPE html>
<html lang="en" data-theme=""><head>
    <title> Xiang Li | Use Python in R </title>

    
    <meta charset="utf-8"><meta name="generator" content="Hugo 0.81.0" /><meta name="viewport" content="width=device-width,initial-scale=1,viewport-fit=cover">
    <meta name="description" content="A lifetime learner. A part of nature.">
    
    <link rel="stylesheet"
          href="https://xiangli2pro.github.io/css/style.min.0c643de4008adca329f7a2d616ce308cca99d4ef45e4cca307323e7857910194.css"
          integrity="sha256-DGQ95ACK3KMp96LWFs4wjMqZ1O9F5MyjBzI&#43;eFeRAZQ="
          crossorigin="anonymous"
          type="text/css">
    
    <link rel="stylesheet"
        href="https://xiangli2pro.github.io/css/markupHighlight.min.f798cbda9aaa38f89eb38be6414bd082cfd71a6780375cbf67b6d2fb2b96491e.css"
        integrity="sha256-95jL2pqqOPies4vmQUvQgs/XGmeAN1y/Z7bS&#43;yuWSR4="
        crossorigin="anonymous"
        type="text/css">
    
        
        
        <link rel="stylesheet"
        href="https://xiangli2pro.github.io/css/customer.min.93e48a477ee532b5600ae5e4aff60bfc28ca5236f93908831b3c74f7423080e1.css"
        integrity="sha256-k&#43;SKR37lMrVgCuXkr/YL/CjKUjb5OQiDGzx090IwgOE="
        crossorigin="anonymous"
        media="screen" />
    
    <link rel="stylesheet"
    href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.1/css/all.min.css"
    integrity="sha512-+4zCK9k+qNFUR5X+cKL9EIR+ZOhtIloNl9GIKS57V1MyNsYpYcUrUeQc9vNfzsWfV28IaLL3i96P9sdNyeRssA=="
    crossorigin="anonymous" />

    
    
      
      
      
      <link rel="preconnect" href="https://fonts.googleapis.com" />
      <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin />
      <link href="https://fonts.googleapis.com/css2?family=Titillium&#43;Web&amp;display=swap" rel="stylesheet" />
    <link rel="shortcut icon" href="https://xiangli2pro.github.io/favicons/favicon.ico" type="image/x-icon">
    <link rel="apple-touch-icon" sizes="180x180" href="https://xiangli2pro.github.io/favicons/apple-touch-icon.png">
    <link rel="icon" type="image/png" sizes="32x32" href="https://xiangli2pro.github.io/favicons/favicon-32x32.png">
    <link rel="icon" type="image/png" sizes="16x16" href="https://xiangli2pro.github.io/favicons/favicon-16x16.png">

    <link rel="canonical" href="https://xiangli2pro.github.io/tech/2021_0409_python-r/">

    
    
    
    
    <script type="text/javascript"
            src="https://xiangli2pro.github.io/js/anatole-header.min.0c05c0a90d28c968a1cad4fb31abd0b8e1264e788ccefed022ae1d3b6f627514.js"
            integrity="sha256-DAXAqQ0oyWihytT7MavQuOEmTniMzv7QIq4dO29idRQ="
            crossorigin="anonymous"></script>


    
    <meta name="twitter:card" content="summary_large_image"/>
<meta name="twitter:image" content="https://xiangli2pro.github.io/images/dandan.jpeg"/>

<meta name="twitter:title" content="Use Python in R"/>
<meta name="twitter:description" content="As a statistician, I use R a lot in my school works and collaboration projects, since R provides almost all the statistical analysis tools I would need in my research, from data wrangling, statistical modeling, simulation, to data visualization."/>


    

</head>
<body><div class="sidebar animated fadeInDown ">
    <div class="logo-title">
        <div class="title">
            <img src="https://xiangli2pro.github.io/images/dandan.jpeg" alt="profile picture">
            <h3 title=""><a href="/">Xiang Li</a></h3>
            <div class="description">
                <p>A lifetime learner. A part of nature.</p>
            </div>
        </div>
    </div>
    <ul class="social-links">
        
            <li>
                <a href="https://www.linkedin.com/in/xiang-li-505757129/" rel="me" aria-label="Linkedin">
                    <i class="fab fa-linkedin fa-2x" aria-hidden="true"></i>
                </a>
            </li>
        
            <li>
                <a href="https://github.com/xiangli2pro" rel="me" aria-label="GitHub">
                    <i class="fab fa-github fa-2x" aria-hidden="true"></i>
                </a>
            </li>
        
    </ul>
    <div class="footer">
        <div class="by_farbox">&copy; Xiang Li  2021-2022 </div>
    </div>
</div>
<div class="main">
    <div class="page-top  animated fadeInDown ">
    <a role="button" class="navbar-burger" data-target="navMenu" aria-label="menu" aria-expanded="false">
        <span aria-hidden="true"></span>
        <span aria-hidden="true"></span>
        <span aria-hidden="true"></span>
    </a>
    <ul class="nav" id="navMenu">
        
        
            
            <li><a 
                   href="/about/"
                        
                   title="">About</a></li>
        
            
            <li><a 
                   href="/tech/"
                        
                   title="">Tech</a></li>
        
            
            <li><a 
                   href="/life/"
                        
                   title="">Life</a></li>
        
            
            <li><a 
                   href="/resource/"
                        
                   title="">Resource</a></li>
        
        
        
    </ul>
</div>

    <div class="autopagerize_page_element">
        <div class="content">
    <div class="post  animated fadeInDown ">
        <div class="post-content">
            
            <div class="post-title">
                <h3>Use Python in R</h3>
                
                    <div class="info">
                        <em class="fas fa-calendar-day"></em>
                        <span class="date"> Fri, Apr 9, 2021
                                           </span>
                        <em class="fas fa-stopwatch"></em>
                        <span class="reading-time">6-minute read</span>
                        
                        <span class="separator"><a class="tag" href="/tags/python/">python</a><a class="tag" href="/tags/r/">r</a><a class="tag" href="/tags/microarray-analysis/">microarray analysis</a></span>

                    </div>
                
            </div>

            <p> </p>
<p>As a statistician, I use R a lot in my school works and collaboration projects, since R provides almost all the statistical analysis tools I would need in my research, from data wrangling, statistical modeling, simulation, to data visualization. Normally statisticians/biostatisticians would employ R to implement and publish their latest statistical techniques, so using R makes it convenient to follow up those up to date developments. However, sometimes I did find out that Python is faster than R when they perform the same task as data size goes larger, for example, when using K-means model and spectral clustering model. There are also many articles online comparing the speed of Python and R in various scenarios. Therefore, it might be a good idea to combine Python and R together to make use of the good parts from both worlds.</p>
<p><a href="https://rstudio.github.io/reticulate/">reticulate</a> is a R package which calls Python from R in a variety of ways including R Markdown, sourcing Python scripts, importing Python modules, and using Python interactively within an R session. In the post, I will present a simple example of microarray analysis, where spectral clustering model is imported from Python to cluster genes into different groups and graphical tools are called in R to display the clustering results. By the end, I will also demonstrate how to transfer data between R and Python.</p>
<h4 id="1-import-data">1. Import data</h4>
<p>I use the gene expression data from a <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75694">study</a> on brewer&rsquo;s yeast cells. The data has 378 samples and 5900 genes and is already normalized<cite><sup id="fnref:1"><a href="#fn:1" class="footnote-ref" role="doc-noteref">1</a></sup></cite>  by the provider. First transform the data into a matrix of 378 rows and 5900 columns, then standardize columns such that each column has mean zero and unit variance.</p>
<div class="highlight"><pre class="chroma"><code class="language-R" data-lang="R"><span class="c1">#load data</span>
<span class="n">gene_df</span> <span class="o">&lt;-</span> <span class="nf">read.delim</span><span class="p">(</span><span class="s">&#34;GSE75694_matrix_378_samples.txt&#34;</span><span class="p">)</span>
<span class="nf">dim</span><span class="p">(</span><span class="n">gene_df</span><span class="p">)</span>
<span class="c1"># 5900  379</span>

<span class="n">gene_names</span> <span class="o">&lt;-</span> <span class="n">gene_df</span><span class="o">$</span><span class="n">ID_REF</span> <span class="o">%&gt;%</span> <span class="nf">as.character</span><span class="p">()</span>
<span class="n">sample_names</span> <span class="o">&lt;-</span> <span class="nf">colnames</span><span class="p">(</span><span class="n">gene_df</span><span class="p">)</span>
<span class="n">gene_mat</span> <span class="o">&lt;-</span> <span class="n">gene_df[</span><span class="p">,</span><span class="m">2</span><span class="o">:</span><span class="nf">ncol</span><span class="p">(</span><span class="n">gene_df</span><span class="p">)</span><span class="n">]</span> <span class="o">%&gt;%</span> <span class="nf">as.matrix</span><span class="p">()</span>
<span class="n">gene_mat</span> <span class="o">&lt;-</span> <span class="nf">t</span><span class="p">(</span><span class="n">gene_mat</span><span class="p">)</span>
<span class="nf">colnames</span><span class="p">(</span><span class="n">gene_mat</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="kc">NULL</span>
<span class="nf">rownames</span><span class="p">(</span><span class="n">gene_mat</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="kc">NULL</span>

<span class="c1"># standardize columns</span>
<span class="n">gene_scale</span> <span class="o">&lt;-</span> <span class="nf">scale</span><span class="p">(</span><span class="n">gene_mat</span><span class="p">)</span> 
</code></pre></div><p> </p>
<h4 id="2-import-python-modules">2. Import Python modules</h4>
<p>One can call Python by either specifying a path or a Conda environment (need to install anaconda). Here I choose to create a Conda environment, and import modules <code>numpy</code> and <code>sklearn.cluster</code>.</p>
<div class="highlight"><pre class="chroma"><code class="language-R" data-lang="R"><span class="c1"># install and load R package reticulate</span>
<span class="nf">install.packages</span><span class="p">(</span><span class="s">&#39;reticulate&#39;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">reticulate</span><span class="p">)</span> 

<span class="c1"># check and create an anaconda working environment</span>
<span class="nf">conda_list</span><span class="p">()</span>
<span class="nf">conda_create</span><span class="p">(</span><span class="s">&#34;r-reticulate&#34;</span><span class="p">)</span>
<span class="nf">use_condaenv</span><span class="p">(</span><span class="s">&#34;r-reticulate&#34;</span><span class="p">,</span> <span class="n">required</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>

<span class="c1"># check python configuration</span>
<span class="nf">py_config</span><span class="p">()</span>

<span class="c1"># install packages</span>
<span class="nf">conda_install</span><span class="p">(</span><span class="s">&#39;r-reticulate&#39;</span><span class="p">,</span> <span class="n">packages</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#39;numpy&#39;</span><span class="p">,</span><span class="s">&#39;pandas&#39;</span><span class="p">),</span> <span class="n">channel</span> <span class="o">=</span> <span class="s">&#39;conda-forge&#39;</span><span class="p">)</span>
<span class="nf">conda_install</span><span class="p">(</span><span class="s">&#39;r-reticulate&#39;</span><span class="p">,</span> <span class="n">packages</span> <span class="o">=</span> <span class="s">&#39;scikit-learn&#39;</span><span class="p">,</span> <span class="n">channel</span> <span class="o">=</span> <span class="s">&#39;conda-forge&#39;</span><span class="p">)</span>

<span class="c1"># import packages</span>
<span class="n">numpy</span> <span class="o">&lt;-</span> <span class="nf">import</span><span class="p">(</span><span class="s">&#39;numpy&#39;</span><span class="p">)</span>
<span class="n">sklearn_cluster</span> <span class="o">&lt;-</span> <span class="nf">import</span><span class="p">(</span><span class="s">&#39;sklearn.cluster&#39;</span><span class="p">)</span>
</code></pre></div><p> </p>
<h4 id="3-spectral-clustering">3. Spectral clustering</h4>
<p>Spectral clustering model  is used to cluster the genes into 20 groups and the group labels will be saved. When fitting the model, the absolute correlation matrix of the standardized genes is set as the affinity matrix and the number of clusters is 20. Note that the indexing system in Python starts from 0, while the indexing system in R starts from 1. Most importantly, when the arguments of Python functions require integer, we need explicitly specify the integer property  in R, i.e. use <strong>&lsquo;20L&rsquo;</strong> instead of &lsquo;20&rsquo;, otherwise, error will occur. After fitting the model, we can take a look at the frequencies of the 20 clustered groups.</p>
<div class="highlight"><pre class="chroma"><code class="language-R" data-lang="R"><span class="c1"># absolute correlation as affinity matrix</span>
<span class="n">gene_scale_abscor</span> <span class="o">&lt;-</span> <span class="nf">r_to_py</span><span class="p">(</span><span class="nf">abs</span><span class="p">(</span><span class="nf">cor</span><span class="p">(</span><span class="n">gene_scale</span><span class="p">)))</span>
<span class="c1"># r_to_py() is not necessary</span>
<span class="c1"># reticulate package will take care of the transformation between R data and Python data</span>

<span class="c1"># cluster 5900 genes into 20 groups</span>
<span class="c1"># !!! specify the integer L</span>
<span class="n">spec_model</span> <span class="o">&lt;-</span> <span class="n">sklearn_cluster</span><span class="o">$</span><span class="nf">SpectralClustering</span><span class="p">(</span><span class="n">n_clusters</span><span class="o">=</span><span class="m">20L</span><span class="p">,</span> <span class="n">affinity</span><span class="o">=</span><span class="s">&#34;precomputed&#34;</span><span class="p">)</span>
<span class="n">spec_fit</span> <span class="o">&lt;-</span> <span class="n">spec_model</span><span class="o">$</span><span class="nf">fit</span><span class="p">(</span><span class="n">X</span><span class="o">=</span><span class="n">gene_scale_abscor</span><span class="p">)</span>

<span class="c1"># save the labels</span>
<span class="n">spec_label</span> <span class="o">&lt;-</span> <span class="n">spec_fit</span><span class="o">$</span><span class="n">labels_</span><span class="m">+1</span>

<span class="c1"># frequency of each group</span>
<span class="nf">table</span><span class="p">(</span><span class="n">spec_label</span><span class="p">)</span>
<span class="c1"># spec_label</span>
<span class="c1">#   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20 </span>
<span class="c1"># 274 203 289 235 392 258 853 356 272 305 211 265 167 397 140 231 364 190 334 164 </span>

</code></pre></div><p> </p>
<h4 id="4-heatmap">4. Heatmap</h4>
<p>In order to visualize the group structure, we display the heatmap of the correlation matrix using the R package <code>ComplexHeatmap</code>. Before making the plot, we define a helper function to rearrange the columns and rows of  the correlation matrix according to their group memberships.</p>
<div class="highlight"><pre class="chroma"><code class="language-R" data-lang="R"><span class="c1"># helper function</span>
<span class="n">reorder_col</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">label</span><span class="p">,</span> <span class="n">K</span><span class="p">){</span>

  <span class="n">order_vec</span> <span class="o">&lt;-</span> <span class="nf">vector</span><span class="p">()</span>
  <span class="nf">for </span><span class="p">(</span><span class="n">k</span> <span class="n">in</span> <span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="n">K</span><span class="p">)){</span>
    <span class="n">order_vec</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="n">order_vec</span><span class="p">,</span> <span class="nf">which</span><span class="p">(</span><span class="n">label</span><span class="o">==</span><span class="n">k</span><span class="p">))</span>
  <span class="p">}</span>

  <span class="nf">return</span><span class="p">(</span><span class="n">order_vec</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># rearrange the columns by groups</span>
<span class="n">col_by_label</span> <span class="o">&lt;-</span> <span class="nf">reorder_col</span><span class="p">(</span><span class="n">spec_label</span><span class="p">,</span> <span class="n">K</span><span class="o">=</span><span class="m">20</span><span class="p">)</span>

<span class="c1"># get correlation matrix</span>
<span class="n">gene_scale_cor</span> <span class="o">&lt;-</span> <span class="nf">cor</span><span class="p">(</span><span class="n">gene_scale</span><span class="p">)</span>
</code></pre></div><div class="highlight"><pre class="chroma"><code class="language-R" data-lang="R"><span class="c1"># install and load package</span>
<span class="nf">install.packages</span><span class="p">(</span><span class="s">&#39;ComplexHeatmap&#39;</span><span class="p">)</span>
<span class="nf">suppressPackageStartupMessages</span><span class="p">(</span><span class="nf">library</span><span class="p">(</span><span class="n">ComplexHeatmap</span><span class="p">))</span>

<span class="c1"># plot heatmap </span>
<span class="nf">Heatmap</span><span class="p">(</span>
  <span class="n">gene_scale_cor</span><span class="p">,</span> 
  
  <span class="n">cluster_rows</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> <span class="n">cluster_columns</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>
  <span class="n">show_row_names</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> <span class="n">show_column_names</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>
  
  <span class="n">column_order</span> <span class="o">=</span> <span class="n">col_by_label</span><span class="p">,</span>
  <span class="n">row_order</span> <span class="o">=</span> <span class="n">col_by_label</span><span class="p">,</span>
  
  <span class="n">show_heatmap_legend</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
  <span class="n">heatmap_legend_param</span> <span class="o">=</span> <span class="nf">list</span><span class="p">(</span>
    <span class="n">title</span> <span class="o">=</span> <span class="s">&#34;gene correlation&#34;</span><span class="p">,</span> 
    <span class="n">title_position</span> <span class="o">=</span> <span class="s">&#34;leftcenter-rot&#34;</span><span class="p">,</span>
    <span class="n">legend_height</span> <span class="o">=</span> <span class="nf">unit</span><span class="p">(</span><span class="m">3</span><span class="p">,</span> <span class="s">&#34;cm&#34;</span><span class="p">)</span>
  <span class="p">),</span>
  
  <span class="n">column_title</span> <span class="o">=</span> <span class="s">&#34;Standardized Gene Expression \n Correlation Matrix K=20&#34;</span><span class="p">,</span>
  <span class="n">use_raster</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
  <span class="n">raster_resize</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> 
  <span class="n">raster_device</span> <span class="o">=</span> <span class="s">&#34;CairoPNG&#34;</span>
<span class="p">)</span>
</code></pre></div><center><img src="/images/blog_corr.png" ></center>
<p>From the heatmap we can observe that the correlation matrix has a clear group structure. Further analysis needs to be conducted to explore the characteristics of the genes of different groups.</p>
<p>Some other clustering models might require the genes to follow certain distribution, for example, the Gaussian mixture model assumes the data follows normal distributions. A  simple visual check on the genes densities can be done by plotting the density heatmap as below. On the left figure, we plot the densities of the first 50 genes in group 1, and on the right figure we compare the densities of 50 columns following the standard normal distribution.</p>
<div class="highlight"><pre class="chroma"><code class="language-R" data-lang="R"><span class="c1"># create a matrix with 50 columns following N(0,1)</span>
<span class="n">rnorm_mat</span> <span class="o">=</span> <span class="nf">matrix</span><span class="p">(</span><span class="nf">rnorm</span><span class="p">(</span><span class="m">378</span><span class="o">*</span><span class="m">50</span><span class="p">),</span> <span class="n">nrows</span><span class="o">=</span><span class="m">378</span><span class="p">)</span>

<span class="c1"># density plot</span>
<span class="n">ht_list</span> <span class="o">=</span> <span class="nf">densityHeatmap</span><span class="p">(</span><span class="n">gene_scale[</span><span class="p">,</span><span class="nf">which</span><span class="p">(</span><span class="n">spec_label</span><span class="o">==</span><span class="m">1</span><span class="p">)</span><span class="n">[1</span><span class="o">:</span><span class="m">50</span><span class="n">]]</span><span class="p">,</span>
                         <span class="n">ylim</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">-2</span><span class="p">,</span><span class="m">2</span><span class="p">),</span>
                         <span class="n">title</span> <span class="o">=</span> <span class="s">&#39;Density of Genes in Group 1 (N=50)&#39;</span><span class="p">,</span>
                         <span class="n">ylab</span> <span class="o">=</span> <span class="s">&#39;Density&#39;</span><span class="p">,</span>
                         <span class="n">heatmap_legend_param</span> <span class="o">=</span> <span class="nf">list</span><span class="p">(</span><span class="n">title</span> <span class="o">=</span> <span class="s">&#34;density&#34;</span><span class="p">)</span>
<span class="p">)</span> <span class="o">+</span> <span class="nf">densityHeatmap</span><span class="p">(</span><span class="nf">scale</span><span class="p">(</span><span class="n">rnorm_mat</span><span class="p">),</span>
                   <span class="n">ylim</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">-2</span><span class="p">,</span><span class="m">2</span><span class="p">),</span>
                   <span class="n">title</span> <span class="o">=</span> <span class="s">&#39;Density of N(0,1)&#39;</span><span class="p">,</span>
                   <span class="n">ylab</span> <span class="o">=</span> <span class="s">&#39;&#39;</span><span class="p">,</span>
                   <span class="n">heatmap_legend_param</span> <span class="o">=</span> <span class="nf">list</span><span class="p">(</span><span class="n">title</span> <span class="o">=</span> <span class="s">&#34;N(0,1) density&#34;</span><span class="p">),</span>
                   <span class="n">show_heatmap_legend</span> <span class="o">=</span> <span class="kc">TRUE</span>
<span class="p">)</span>

<span class="nf">draw</span><span class="p">(</span><span class="n">ht_list</span><span class="p">)</span>
</code></pre></div><center><img src="/images/density.png"></center>
<p> </p>
<h4 id="5-transfer-data-between-python-and-r">5. Transfer data between Python and R</h4>
<p>Package <code>reticulate</code> will convert R data type to the corresponding Python data type and vice versa in R environment, but sometimes one may still need to transfer data between R and Python. In this case, we can use <code>feather</code> package. For example, we save the clustering labels <code>spec_label</code> in R then load it in Python.</p>
<div class="highlight"><pre class="chroma"><code class="language-R" data-lang="R"><span class="c1">## save data in R as .feather</span>
<span class="nf">install.packages</span><span class="p">(</span><span class="s">&#39;feather&#39;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="s">&#39;feather&#39;</span><span class="p">)</span>

<span class="nf">write_feather</span><span class="p">(</span><span class="n">spec_label</span><span class="p">,</span> <span class="s">&#34;spec_label_R.feather&#34;</span><span class="p">)</span>
</code></pre></div><div class="highlight"><pre class="chroma"><code class="language-Python" data-lang="Python"><span class="c1">## load R data in Python</span>
<span class="kn">import</span> <span class="nn">feather</span>

<span class="n">spec_label_Python</span> <span class="o">=</span> <span class="n">feather</span><span class="o">.</span><span class="n">read_dataframe</span><span class="p">(</span><span class="s2">&#34;spec_label_R.feather&#34;</span><span class="p">)</span>
<span class="n">feather</span><span class="o">.</span><span class="n">write_dataframe</span><span class="p">(</span><span class="n">spec_label_Python</span><span class="p">,</span> <span class="s2">&#34;spec_label_Python.feather&#34;</span><span class="p">)</span>
</code></pre></div><div class="highlight"><pre class="chroma"><code class="language-R" data-lang="R"><span class="c1">## load Python data in R</span>

<span class="nf">read_feather</span><span class="p">(</span><span class="s">&#34;spec_label_Python.feather&#34;</span><span class="p">)</span>
</code></pre></div><p> 
 </p>
<section class="footnotes" role="doc-endnotes">
<hr>
<ol>
<li id="fn:1" role="doc-endnote">
<p>Data is normalized using dChip implementation with default parameters from the affy package (1.32.1) within R/Bioconductor. For beginners to the microarray analysis, one can check <a href="https://www.mrc-lmb.cam.ac.uk/genomes/madanm/microarray/chapter-final.pdf">An Introduction to Microarray Data Analysis</a> to learn the basic concepts of microarray data. <a href="#fnref:1" class="footnote-backref" role="doc-backlink">&#x21a9;&#xfe0e;</a></p>
</li>
</ol>
</section>
</div>
        <div class="post-footer">
            <div class="info">
                
                
            </div>
        </div>

        <div id="fb_comments_container">
              
                <h2></h2>
                <script src="https://utteranc.es/client.js"
        repo="xiangli2pro/xiangli2pro.github.io"
        issue-term="pathname"
        theme="github-light"
        
        crossorigin="anonymous"
        async>
</script>

            </div>
    </div>


        </div>
    </div>
</div>

<script type="text/javascript"
        src="https://xiangli2pro.github.io/js/jquery.min.64d0083866906099f942140bc1c5cba4b1bf65783c52e4a63c5c46bf9dbc41f4.js"
        integrity="sha256-ZNAIOGaQYJn5QhQLwcXLpLG/ZXg8UuSmPFxGv528QfQ="
        crossorigin="anonymous"></script>




<script type="text/javascript"
        src="https://xiangli2pro.github.io/js/bundle.min.45159b0e20ba3878972e064b7edc464c31a9e35a0d0a6a71e3fec84efdbe2ea5.js"
        integrity="sha256-RRWbDiC6OHiXLgZLftxGTDGp41oNCmpx4/7ITv2&#43;LqU="
        crossorigin="anonymous"></script>

<script type="text/javascript"
        src="https://xiangli2pro.github.io/js/medium-zoom.min.2d6fd0be87fa98f0c9b4dc2536b312bbca48757f584f6ea1f394abc9bcc38fbc.js"
        integrity="sha256-LW/Qvof6mPDJtNwlNrMSu8pIdX9YT26h85SrybzDj7w="
        crossorigin="anonymous"></script><script defer
                type="text/javascript"
                src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/MathJax.js?config=TeX-AMS-MML_HTMLorMML"
                integrity="sha384-e/4/LvThKH1gwzXhdbY2AsjR3rm7LHWyhIG5C0jiRfn8AN2eTN5ILeztWw0H9jmN"
                crossorigin="anonymous"></script>
        <script type="text/x-mathjax-config">
            MathJax.Hub.Config({ tex2jax: { inlineMath: [['$','$'], ['\\(','\\)']] } });
        </script></body>

</html>
